### PR TITLE
Ticket/15756 Add Lost Username function

### PR DESCRIPTION
**Pull request by v12mike**
Checklist:

- [ ] Correct branch: master for new features; 3.2.x for fixes
- [ ] Tests pass
- [ ] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html) and [3.2.x](https://area51.phpbb.com/docs/dev/3.2.x/development/coding_guidelines.html)
- [ ] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.2.x/development/git.html)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-15756

### CODE REVIEW

**Comment by v12mike** [comment]
ticket/15756 has been rebased so that it does not include ticket 15726.

Regarding the pull request to 3.2.x rather than master, I consider this functionality to be an extension of the existing 'forgot password' feature, the new code is closely based on that.  I hoope that can be the consensus view.

**Comment by v12mike** [comment]
I have set up a test board at: https://xantiav6.co.uk/forgotusername/  you can create an account there and try retrieving your forgotten username.

That board is configured to allow email address re-use, so you can verify that case also.

**Comment by v12mike** [comment]
reopened due to (random?) Travis failure

**Comment by rubencm** [comment]
Why don't just require email and send user and new password?

**Comment by v12mike** [comment]
I think that it is more logical to deal with one item (username or password) at a time and leave it to the user to decide if they need a new password.  It would also get a bit ugly in the case where there are multiple usernames per email address.

File: phpBB/language/en/email/user_remind_username.txt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+Subject: {SITENAME} Username reminder
+
+Hello {USERNAME},
+
+Somebody (perhaps you) has requested a reminder of the username for your account at "{SITENAME}". Your Username is: {USERNAME}
+
+You can login to "{SITENAME}"at {U_LOGIN}, or use the 'login' link on the forum home page: .  Note that the username and password are case-sensitive.
```
Comment (lavigor) [inline phpBB/language/en/email/user_remind_username.txt]: This line has several issues:
* no space between site name and `at`;
* empty homepage link.

Comment (v12mike) [reply phpBB/language/en/email/user_remind_username.txt]: I have pushed a fix.  

Note that there is intentionally no homepage link, as the text  'or use the 'login' link on the forum home page.'  is an instruction intended for users who prefer not to click links in emails.

**Comment by Elsensee** [comment]
Wouldn't it be more user-friendly if we just send one email with all usernames instead of spamming the users?
Also.. maybe we need a captcha on that page I guess?

**Comment by Nicofuma** [comment]
Thanks for your contribution but it should target 3.3 (master). Could you rebase?

File: phpBB/includes/ucp/ucp_username.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+					'U_LOGIN'	=> "$server_url/ucp.$phpEx?mode=login",
+					'U_FORGOT_PASS'	=> "$server_url/ucp.$phpEx?mode=sendpassword")
+				);
+				$messenger->send($user_row['user_notify_type']);
+			}
+			$db->sql_freeresult($result);
+
+			if ($count == 0)
+			{
+				trigger_error('NO_EMAIL_USER');
```
Comment (Nicofuma) [inline phpBB/includes/ucp/ucp_username.php]: I wouldn't trigger an error if the email is not used

File: phpBB/includes/ucp/ucp_username.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+			}
+
+			meta_refresh(3, append_sid("{$phpbb_root_path}index.$phpEx"));
+
+			$message = $user->lang(['USERNAME_REMINDER_SENT'], $count) . '<br /><br />' . sprintf($user->lang['RETURN_INDEX'], '<a href="' . append_sid("{$phpbb_root_path}index.$phpEx") . '">', '</a>');
```
Comment (Nicofuma) [inline phpBB/includes/ucp/ucp_username.php]: I won't say the number of ent emails either

File: phpBB/language/en/email/user_remind_username.txt [hunk 4]
```diff
@@ -0,0 +1,13 @@
+Subject: {SITENAME} Username reminder
+
+Hello {USERNAME},
+
+Somebody (perhaps you) has requested a reminder of the username for your account at "{SITENAME}". Your Username is: {USERNAME}
```
Comment (Nicofuma) [inline phpBB/language/en/email/user_remind_username.txt]: I would say that someone requested a reminder of the username for accounts using the email address XXX personnaly

File: phpBB/includes/ucp/ucp_username.php (hunk 2 above)
Comment (v12mike) [reply phpBB/includes/ucp/ucp_username.php]: OK

File: phpBB/includes/ucp/ucp_username.php (hunk 3 above)
Comment (v12mike) [reply phpBB/includes/ucp/ucp_username.php]: OK

File: phpBB/language/en/email/user_remind_username.txt (hunk 4 above)
Comment (v12mike) [reply phpBB/language/en/email/user_remind_username.txt]: OK

File: phpBB/includes/ucp/ucp_username.php (hunk 3 above)
Comment (v12mike) [reply phpBB/includes/ucp/ucp_username.php]: I will make the suggested changes and rebase to the master branch.

**Pull request closed without merging**